app/lib/task/backend.dart

```diff
@@ -1169,6 +1161,41 @@ String? _extractBearerToken(shelf.Request request) {
   return parts.last.trim();
 }
 
+PackageVersionStateInfo _extractAndVerifyVersionState(
+  String package,
+  String version,
+  PackageState? state,
+  shelf.Request request,
+) {
+  final token = _extractBearerToken(request);
+  if (token == null) {
+    throw AuthenticationException.authenticationRequired();
+  }
+  if (state == null) {
+    throw NotFoundException.resource('$package/$version');
+  }
+  final versionState = state.versions![version];
+  if (versionState == null) {
+    // check if the task was aborted
+    final abortedToken =
+        state.abortedTokens?.firstWhereOrNull((t) => t.token == token);
+    if (abortedToken != null && abortedToken.expires.isBefore(clock.now())) {
+      throw TaskAbortedException('$package/$version has been aborted.');
+    }
+    // otherwise throw a generic not found error
+    throw NotFoundException.resource('$package/$version');
+  }
+
+  // Check the secret token
+  if (!versionState.isAuthorized(token)) {
+    throw AuthenticationException.authenticationRequired();
+  }
```

Comment:
This is vulnerable to timing attacks!

I'd suggest adding an `isAuthorized` method to `AbortedTokenInfo` object.

Also `versionState != null` doesn't imply that the token couldn't be aborted.

I'd suggest:
 * Always check `abortedTokens`, ideally always compare all of them (not just the first, because we want fixed time behavior)
 * Then check `versionState.isAuthorized`

Reply:
Done.